src/core/utils/components/VirtualScroll.tsx

```diff
@@ -0,0 +1,229 @@
+import React from 'react';
+import { CommonProps } from '../props';
+
+const getScrollableParent = (element: HTMLElement | null): HTMLElement => {
+  if (!element || element === document.body) {
+    return document.body;
+  }
+
+  return isElementScrollable(element)
+    ? element
+    : getScrollableParent(element.parentElement);
+};
+
+const isElementScrollable = (element: HTMLElement) => {
+  return /(auto|scroll|overlay)/.test(
+    getElementStyle(element, 'overflow') +
+      getElementStyle(element, 'overflow-y'),
+  );
+};
+
+const getElementStyle = (element: HTMLElement, prop: string) => {
+  return getComputedStyle(element, null).getPropertyValue(prop);
+};
+
+const getElementHeight = (element: HTMLElement | undefined) => {
+  return element?.getBoundingClientRect().height ?? 0;
+};
+
+const getNumberOfNodesInHeight = (childHeight: number, totalHeight: number) => {
+  if (!childHeight) {
+    return 0;
+  }
+
+  return Math.floor(totalHeight / childHeight);
+};
+
+const getTranslateValue = (childHeight: number, startIndex: number) => {
+  return childHeight * startIndex;
+};
+
+const getVisibleNodeCount = (
+  childHeight: number,
+  startIndex: number,
+  childrenLength: number,
+  scrollContainer: HTMLElement,
+) => {
+  return Math.min(
+    childrenLength - startIndex,
+    getNumberOfNodesInHeight(childHeight, getElementHeight(scrollContainer)),
+  );
+};
+
+export type VirtualScrollProps = {
+  /**
+   * Length of the items to virtualize.
+   */
+  itemsLength: number;
+  /**
+   * Single item render function, which gives index of the item (0 based) in the data array
+   * and expects to get the JSX of that element to render.
+   * Recommended to memoize the reference of the function.
+   */
+  itemRenderer: (index: number) => JSX.Element;
+  /**
+   * Number of items to be rendered at the end.
+   * Not recommended to go lower than the visible items in viewport.
+   * @default 20
+   */
+  bufferSize?: number;
+} & Omit<CommonProps, 'title'>;
+
+/**
+ * `VirtualScroll` component used to render a huge amount of items in the DOM. It renders only the ones which are visible
+ * and the amount provided through `bufferSize` prop. Can be used inside other components like `Table`.
+ *
+ * It has two wrapper elements, so DOM will be changed. One is used for setting full expected height in the scrollable container
+ * and other is for transformation (translateY) to show the correct part of the list.
+ *
+ * Currently it works only with the direct scrollable parent element. It does not work with body scroll.
+ * It supports only static (same) height rows virtualization. Expect some issues, if list consists of different height elements.
+ * @example
+ * const itemRenderer = React.useCallback(() => (
+ *  <div key={index}>
+ *    This is my item #{index}
+ *  </div>
+ * ), [])
+ * <VirtualScroll
+ *  itemsLength={1000}
+ *  itemRenderer={itemRenderer}
+ * />
+ */
+export const VirtualScroll = ({
```

Comment:
Should we mark it with `@beta`? Just in case something needs to be changed? Or not make it public and wait until we apply it to `ComboBox` and `Tree` to see if everything is supported.

Reply:
So I just removed from core barrel, so its like an inner component for now.

Reply:
I think we use `@private` in some of our hooks to explicitly say it's internal.

Reply:
And the `@beta` flag we might still want to add in the table prop